### PR TITLE
PoC implementing a cache for results of checks of "svnauthz accessof".

### PR DESCRIPTION
**Pull request by ams-tschoening**
#78 revealed that access checks are at least performed twice for at least the index page because of creating the project selection form, so I tried to work around that by checking if the current request belongs to the index page and not create that form anymore. The problem is that WebSVN can work with MultiViews vs. not, in which case the index page is different and MultiViews could have been setup wrongly, like I did in my tests, resulting in an additional state which needs to be checked.

That doesn't make much sense in my opinion so I implemented a PoC of a cache for results of permissions checks. While such caches are somewhat unsafe, it's fully transparent to all callers, should only last about the request itself and avoids the need to deal with different entry points under different circumstances altogether.

Additionally, if e.g. a project selection form should be shown or not should be up to templates in theory and not decided based on the request to some index page. So the former approach is pretty much a workaround as well.

So let's take this PR to discuss things.

Especially because I have the feeling there are some additional bottlenecks like this one: Loading the index page took 55 s before this PR, 35 afterwards. Loading an individual repo takes ~20 s, most likely because of the project selection form checking access to all repos. But if that takes ~20 s only for one individual repo, why is it slower for the index page?

I guess it's because of things like checking all repos for `db`-files and such, but someone needs to look at this at some point.

### CODE REVIEW

**Comment by michael-o** [comment]
Codewise it looks fine to me, one might to enable the cache explicitly and not implicitly. I will test this on the server too.

File: include/authz.php [hunk 1]
```diff
@@ -61,26 +62,58 @@ function setUsername() {
 	// Private function to simplify creation of common SVN authz command string text.
 	function svnAuthzCommandString($repos, $path, $checkSubDirs = false) {
 		global $config;
-		return $config->getSvnAuthzCommand().' --repository '.$repos.' --path '.quote($path).
-			' '.($this->hasUsername() ? '--username '.quote($this->user).' ' : '').($checkSubDirs ? '-R ' : '').quote($this->accessfile);
+
+		$cmd			= $config->getSvnAuthzCommand();
+		$repoAndPath	= '--repository ' . quote($repo) . ' --path ' . quote($path);
```
Comment (michael-o) [inline include/authz.php]: Here is a severe issue, has to be `$repos`

Comment (ams-tschoening) [reply include/authz.php]: It has to be `$repo` in the function signature instead, must have C&Ped something instead of cherry-picked.

**Comment by ams-tschoening** [comment]
> [...]one might to enable the cache explicitly and not implicitly.

I didn't want to introduce another setting by purpose, as that needs to be documented and we already know of the performance problems and have no solution yet without the cache. One might discuss enabling it on Windows only at all? But in theory other OSes should benefit as well.

> I see no difference when curl gets this main page.

But you are using an authz-file at all with WebSVN? I just retested and see a significant difference of ~20 s between using the cache and not using it. While I have ~120 repos at the index page, I would have expected at least some difference for you as well.

> Especially because I have the feeling there are some additional bottlenecks like this one: Loading the index page took 55 s before this PR, 35 afterwards. Loading an individual repo takes ~20 s, most likely because of the project selection form checking access to all repos. But if that takes ~20 s only for one individual repo, why is it slower for the index page?

I have additional hints: With the current implementation, `svnauthz` consumes ~13 s of ~35 s total processing time. Without the cache it seems to be roughly the double, which makes sense because all repos are simply checked twice. Pretty soon afterwards the following is repeated for each and every repo on the index page and consumes ~17 s in total:

    "C:/Program Files/CollabNet/Subversion Client\svn" --non-interactive --config-dir /tmp/websvn log --xml --quiet --limit 1 "file:///C:/Users/[...]/Src/Alda/@"

**Comment by ams-tschoening** [comment]
Just came across the following, where it's discussed to provide a cache for authz-results for SVN itself as well:

> The patch augments the svn_authz_t struct wih a field "cache" which
is a apr_hash_t, mapping paths to already known allow/deny information.

https://mail-archives.apache.org/mod_mbox/subversion-dev/201308.mbox/<CAC9r9zbkTYMu07F4Zn3arvva-AKiSLoW6MniZVhv0bsTDZ0XmA@mail.gmail.com>

Not sure what the results were, though.

**Comment by michael-o** [comment]
The links seems to be broken. Even if, how will this help? We are starting a process and it exits. To cache will help here.

**Comment by ams-tschoening** [comment]
Fixed the links, the Apache list archives always introduce `<` and `>` for some reason. I only documented this because it might be of interest for a general discussion about using a cache or not, not because it might help technically.

**Comment by michael-o** [comment]
I think the patch only works because the module is loaded into memory and held in memory in some `apr_pool`. We don't have something like this, do we?

**Comment by ams-tschoening** [comment]
Yes, no. :-) The linked approach only caches per request/connection anyway as well, not long-term. The point really only was that the SVN-devs didn't say "NEVER!!!1!11" right from the start.

**Pull request merged**